typescript/raven.d.ts

```diff
@@ -232,6 +307,81 @@ interface RavenTransportOptions {
     onFailure: () => void;
 }
 
-interface RavenPlugin {
+export interface RavenReportDialogOptions {
+    eventId?: number;
+    dsn?: string;
+    user?: {
+        name?: string;
+        email?: string;
+    };
+}
+
+
+// --- Helper Interfaces for complex Data Structures --------------
+export interface RavenPlugin {
     (raven: RavenStatic, ...args: any[]): RavenStatic;
 }
+
+export interface RavenUserContext {
+    id?: string;
+    username?: string;
+    email?: string;
+    ip_address?: string;
+    extra?: { [prop: string]: any };
+}
+
+export interface RavenBreadcrumb {
+    message: string;
+    data: { [id: string]: string };
+    category: string;
+    level: string;
+    timestamp: number;
+}
+
+export interface RavenOutgoingData {
+    /* either message + stacktrace or exception + culprit */
+    message?: string;
+    stacktrace?: {
+        frames: [RavenStacktraceFrame]
+    };
+    exception?: {
+        values: [{
+            type: string;
+            value: string;
+            stacktrace: {
+                frames: [RavenStacktraceFrame];
+            };
+        }];
+    };
+    culprit?: string;
+    request?: {  /* maybe not included */
+        headers: {
+            'User-Agent': string;
+            Referer: string;
+        };
+        url: string;
+    };
+    breadcrumbs: {
+        values: [RavenBreadcrumb];
+    };
+    user: RavenUserContext;
+    level: string;
+    environment: string;
+    release: string;
+    server_name: string;
+    project: string;
+    logger: string;
+    platform: string;
+    tags: { [id: string]: string };
+    extra: { [prop: string]: any };
+    event_id: string;
+    fingerprint: string;
+}
+
+export interface RavenStacktraceFrame {
+     filename: string;
+     lineno:number;
+     colno: string;
+     function:string;
+     in_app: boolean;
+}
```

Comment:
At 387 lines, this definition file is almost as complicated as Raven.js itself. Is it normal for libraries to expose something so rigid? I'm terrified this file will become really painful to maintain.